### DISCUSSION TITLE
Proposal: allow contravariant out parameters

### DISCUSSION THREAD

**Discussion by YairHalberstadt** [dotnet/csharplang#1760] [Language Ideas]
Given types A and B where A : B
I suggest the following code should be valid

```
// Declare a method using an A out parameter
public static Method(out A a) => new a();

...

// Call the method with a B
Method(out B b);

//This should be compiled to
Method(out A temp);
B b = temp;
```

A use of this might be passing an IReadOnlyList as an out parameter to something which accepts a List.

**Comment by HaloFour** [comment]
Are you suggesting this for both regular `out` parameters and out declarations?

```cs
// given
void Method(out List<string> l) {
    l = new List<string>();
}

// regular out parameter
IReadOnlyList<string> l1;
Method(out l1);

// out declaration
Method(out IReadOnlyList<string> l2);
```

**Comment by YairHalberstadt** [comment]
Yes, by creating a temporary variable of the correct type

**Comment by HaloFour** [comment]
That would change the behavior of `out` parameters in subtle ways.  When the callee assigns the `out`/`ref` parameter the target of that reference is immediately assigned, which is observable.  With a temporary variable the target can't be assigned until after the callee returns.

For example:

```cs
public class Foo {
    private IReadOnlyList<string> r;
    private List<string> l;

    public void M(out List<string> list) {
        list = new List<string>();
        throw new Exception("oops!");
    }

    public void Test1() {
        try {
            M(out l);
        }
        catch { }
        Debug.Assert(l != null); // succeeds
    }

    public void Test2() {
        try {
            M(out r);
        }
        catch { }
        Debug.Assert(r != null); // fails
    }
}

```

**Comment by Joe4evr** [comment]
`Test2` doesn't even compile, [according to SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8kAmIA1AD4CuUmAdgOYAEAyrrHGANwCwAUJTQ81ZgAdAGEA9gBtJcAMYxM46lGEBxONSyyuvKnSYt4IgCKYItauNiZZUHTwAO5YJJv1Zk6FHoAxceJ4Abx56UPoHBEw0CHh6AEkAJTgIAgB5aklcABlMWAAeWEi6AD56BB0w8Mjo2Jz8wv5SyXtuSqcXNzRxTAJ6AFkACnFyGHo6mAKYItom3JgAShCw4NbKsNdYegBeek0AdzG5yenigfmKtdCYAAsEcQP9+gBRAA9ZOAcFJQGAIn8HKAAQh+5yWoQAvjwweFnK5ZPQuj16AAVOCwACMZ2hK0uVwQuGhlRxuMqg2Go0koNWl0h1LWshismu9EC9FpJOMcGA5FowgAglAoIgYANJPRATtqORpOd6AB6OX0KDkWTvOAEKDQ2nQ9pwhHdXqo2AAJixdOJlymBLpRMJuLJIzKVJJ7NxDJgTJZbLtoU53N5AqFCBFCHFkullPY8sV2AgmEkmrptPBQA=).

**Comment by HaloFour** [comment]
@Joe4evr

Not today.  This proposal would seek to change that.

**Comment by YairHalberstadt** [comment]
@HaloFour

I presume the original semantics could be preserved by the compiler wrapping it in a try finally block.

So given your example

```
public class Foo {
    private IReadOnlyList<string> r;
    private List<string> l;

    public void M(out List<string> list) {
        list = new List<string>();
        throw new Exception("oops!");
    }

    public void Test1() {
        try {
            M(out l);
        }
        catch { }
        Debug.Assert(l != null); // succeeds
    }

    public void Test2() {
        try {
            M(out r);
            /* This compiles down to
            try { M(out var temp); }
            finally { r = temp; } */
        }
        catch { }
        Debug.Assert(r != null); // succeeds
    }
}
```

**Comment by HaloFour** [comment]
@YairHalberstadt 

That's only one of the scenarios in which it's observable.  Here is another:

```cs
public class C {
    private IReadOnlyList<string> r;

    public void M(out List<string> list) {
        list = new List<string>();
        M2();
    }

    public void M2() {
        Debug.Assert(list != null);
    }

    public void Test() {
        M(out r);
    }
}
```

These may be minor points.  But at the same time the amount of syntax to workaround this intentional limitation is not onerous.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/issues/317#issuecomment-288668177

**Comment by YairHalberstadt** [comment]
@HaloFour 
I Agree. I don't think I appreciated the shear complexity even such a minor addition throws up.

**Comment by YairHalberstadt** [comment]
Still I think I'll leave this open in case someone comes up with a better way of achieving something similar?

**Comment by HaloFour** [comment]
Sure.  I'm not necessarily arguing against the idea, just noting that it would behave differently.  The same would be true if C# were to move to [support `ref`/`out` properties](https://github.com/dotnet/csharplang/issues/748).  The compiler would have to emit IL to invoke the property setter after the method returned (probably in a `finally` block) so the change to the reference wouldn't be immediately available.  But I think it's a good idea to at least get all of these little details on the table so that they can be discussed accordingly.

**Comment by jnm2** [comment]
I want this outcome regardless of the method used to achieve it.

To be fair, if your program depends on an out parameter being observable before the method returns, you probably have bigger issues. :D
